service/webhook/main.go

```diff
@@ -0,0 +1,105 @@
+	//+kubebuilder:scaffold:imports
+)
+
+const (
+	desiredSubscriptionChannelAnnotationKey = "ocs.openshift.io/subscription.channel"
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+func init() {
+	ctrlLog.SetLogger(zap.New())
+	utilruntime.Must(v1alpha1.AddToScheme(scheme))
+	//+kubebuilder:scaffold:scheme
+}
+
+func main() {
+	setupLog := ctrl.Log.WithName("setup")
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:  scheme,
+		Metrics: metrics.Options{BindAddress: "0"},
+		WebhookServer: webhook.NewServer(webhook.Options{
+			Port:    9443,
+			CertDir: "/etc/tls/private",
+		}),
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to create manager")
+		os.Exit(1)
+	}
+	//+kubebuilder:scaffold:builder
+
+	hookServer := mgr.GetWebhookServer()
+	setupLog.Info("registering Subscription Channel validating webhook endpoint")
+	hookServer.Register("/validate-subscription", &webhook.Admission{
+		Handler: &SubscriptionAdmission{
+			Client:  mgr.GetClient(),
+			Decoder: admission.NewDecoder(mgr.GetScheme()),
+			Log:     ctrl.Log.WithName("webhook.subscription"),
+		}},
+	)
+
+	setupLog.Info("starting manager")
+	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+		setupLog.Error(err, "failed to start manager")
+		os.Exit(1)
+	}
+}
+
+type SubscriptionAdmission struct {
+	Client  client.Client
+	Decoder *admission.Decoder
+	Log     logr.Logger
+}
+
+func (s *SubscriptionAdmission) Handle(ctx context.Context, req admission.Request) admission.Response {
+	s.Log.Info("Request received for admission review")
+	subscription := &opv1a1.Subscription{}
+	if err := s.Decoder.Decode(req, subscription); err != nil {
+		s.Log.Error(err, "failed to decode admission reviews as subscription")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("only subscriptions admission reviews are supported: %v", err))
+	}
+
+	if subscription.Spec.Package != "ocs-client-operator" {
+		s.Log.Info("subscription package is not 'ocs-client-operator'")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("only ocs-client-operator subscription validation is supported"))
+	}
+
+	storageClients := &v1alpha1.StorageClientList{}
+	if err := s.Client.List(ctx, storageClients); err != nil {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed to list storageclients for validating subscription request: %v", err))
+	}
+
+	for idx := range storageClients.Items {
+		storageClient := &storageClients.Items[idx]
+		if storageClient.GetAnnotations()[desiredSubscriptionChannelAnnotationKey] != subscription.Spec.Channel {
+			s.Log.Info(fmt.Sprintf("Rejecting review as it doesn't conform to storageclient %q desired subscription channel", client.ObjectKeyFromObject(storageClient)))
+			return admission.Denied(fmt.Sprintf("subscription channel %q not allowed as it'll violate storageclient %q requirements", subscription.Spec.Channel, client.ObjectKeyFromObject(storageClient)))
+		}
+	}
```

Comment:
I believe this code is wrong. Given multiple clients, you cannot assume all of them will indicate the same channel name this means that if you have 2 clients that indicate different channels then you will always reject and the entire system is stuck. 

The correct approach is to find the channel, between all desired channels, that indicate the lowest version and allow the upgrade to that channel.

Reply:
as discussed, the intention and code is correct, however renamed vars to be more accurate.